plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/query/DynamicTableBuilder.java

```diff
@@ -142,41 +155,55 @@ private static Type toTrinoType(DataSchema.ColumnDataType columnDataType)
+                    break;
+                case IDENTIFIER:
+                    if (expressionContext.getIdentifier().equals(WILDCARD)) {
+                        pinotColumnNamesBuilder.addAll(columnHandles.values().stream().map(handle -> ((PinotColumnHandle) handle).getColumnName()).collect(toImmutableList()));
+                    }
+                    else {
+                        pinotColumnNamesBuilder.add(getColumnHandle(schemaTableName, expressionContext.getIdentifier(), columnHandles).getColumnName());
+                    }
+                    break;
+                case FUNCTION:
+                    ExpressionContext rewrittenContext = rewriteExpressionContext(schemaTableName, expressionContext, columnHandles);
+                    pinotColumnNamesBuilder.add(rewrittenContext.getFunction().toString());
+                    break;
+                default:
+                    throw new PinotException(PINOT_UNSUPPORTED_COLUMN_TYPE, Optional.empty(), "Unsupported pinot expression type: " + expressionContext.getType());
             }
         }
         return pinotColumnNamesBuilder.build();
     }
 
-    private static OptionalLong getTopNOrLimit(BrokerRequest request)
+    private static ExpressionContext rewriteExpressionContext(SchemaTableName schemaTableName, ExpressionContext expressionContext, Map<String, ColumnHandle> columnHandles)
     {
-        if (request.getGroupBy() != null) {
-            return OptionalLong.of(request.getGroupBy().getTopN());
-        }
-        else if (request.getSelections() != null) {
-            return OptionalLong.of(request.getSelections().getSize());
-        }
-        else {
-            return OptionalLong.empty();
+        switch (expressionContext.getType()) {
+            case LITERAL:
+                return expressionContext;
+            case IDENTIFIER:
+                String pinotColumnName = getColumnHandle(schemaTableName, expressionContext.getIdentifier(), columnHandles).getColumnName();
+                return ExpressionContext.forIdentifier(pinotColumnName);
+            case FUNCTION:
+                List<ExpressionContext> arguments = expressionContext.getFunction().getArguments().stream()
+                        .map(context -> rewriteExpressionContext(schemaTableName, context, columnHandles))
+                        .collect(toImmutableList());
+                return ExpressionContext.forFunction(new FunctionContext(expressionContext.getFunction().getType(), expressionContext.getFunction().getFunctionName(), arguments));
+            default:
+                throw new PinotException(PINOT_UNSUPPORTED_COLUMN_TYPE, Optional.empty(), "Unsupported pinot expression type: " + expressionContext.getType());
         }
     }
 
-    private static OptionalLong getOffset(BrokerRequest request)
+    private static OptionalLong getOffset(QueryContext queryContext)
     {
-        if (request.getSelections() != null && request.getSelections().getOffset() > 0) {
```

Comment:
Instead of removing the check let's add a `checkState/verify` here since there may be other callers of this method.

Or does the Pinot library populate offset correctly even when no selections are available?

Reply:
In pinot-0.8.0 all the request.get# methods return null. I moved them to use the QueryContext api.